File: compiler/internal/vm/static_core/plugins/ets/tests/interop_js/tests/generic/ts_to_ets/generic.ts
/*
 * Copyright (c) NeXTHub Corporation. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Author: Tunjay Akbarli
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
 * Middletown, DE 19709, New Castle County, USA.
 */

export const tsString = 'string';
export const tsNumber = 1;
type UnionType = string | number;

type LiteralType = 1 | 'string';

export class LiteralClass<T extends LiteralType> {
	private _value: T;

	constructor(value: T) {
		this._value = value;
	}

	set(arg: T): void {
		this._value = arg;
	}

	get(): T {
		return this._value;
	}
}

export class UnionClass<T extends UnionType> {
	private _value: T;

	constructor(value: T) {
		this._value = value;
	}

	set(arg: T): void {
		this._value = arg;
	}

	get(): T {
		return this._value;
	}
}

interface GInterface<T> {
	value: T;
}

export class InterfaceClass<T> implements GInterface<T> {
	public value: T;

	constructor(value: T) {
		this.value = value;
	}

	set(arg: T): void {
		this.value = arg;
	}

	get(): T {
		return this.value;
	}
}

export abstract class GAbstract<T> {
	private _value: T;

	constructor(value: T) {
		this._value = value;
	}

	set(arg: T): void {
		this._value = arg;
	}

	get(): T {
		return this._value;
	}
}

export class AbstractClass<T> extends GAbstract<T> {
	constructor(value: T) {
		super(value);
	}
}

export class GClass<T> {
	public content: T;

	constructor(content: T) {
		this.content = content;
	}

	public get(): T {
		return this.content;
	}
}

export function genericFunction<T>(arg: T): T {
	return arg;
}

export function tupleDeclaredType<T, U>(items: [T, U]): [T, U] {
	return items;
}

interface Data {
	data: string;
}

export function genericSubsetRef<T extends Data>(items: T): T {
	return items;
}

type TGenericFn<T> = () => T;

export const explicitlyDeclaredType: TGenericFn<string> = () => {
	return tsString;
};

export const literalClass = new LiteralClass(tsString);
export const unionClass = new UnionClass<string>(tsString);
export const interfaceClass = new InterfaceClass<string>(tsString);
export const abstractClass = new AbstractClass<string>(tsString);

File: compiler/internal/vm/static_core/plugins/ets/tests/interop_js/tests/generic/ts_to_ets/generic.cpp
/*
 * Copyright (c) NeXTHub Corporation. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Author: Tunjay Akbarli
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
 * Middletown, DE 19709, New Castle County, USA.
 */

#include <gtest/gtest.h>
#include "ets_interop_js_gtest.h"

namespace ark::ets::interop::js::testing {

class EtsGenericTsToEtsTest : public EtsInteropTest {};

TEST_F(EtsGenericTsToEtsTest, check_literal_class_generic)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkLiteralClassGeneric"));
}

TEST_F(EtsGenericTsToEtsTest, check_union_class_generic)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkUnionClassGeneric"));
}

TEST_F(EtsGenericTsToEtsTest, check_interface_class_generic)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkInterfaceClassGeneric"));
}

TEST_F(EtsGenericTsToEtsTest, check_abstract_class_generic)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkAbstractClassGeneric"));
}

TEST_F(EtsGenericTsToEtsTest, check_generic_class_generic)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericClassGeneric"));
}

TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_literal_class_object_form_ts)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericLiteralClassObjectFromTs"));
}

TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_union_class_object_form_ts)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericUnionClassObjectFromTs"));
}

TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_interface_class_object_form_ts)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericInterfaceClassObjectFromTs"));
}

TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_abstract_class_object_form_ts)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericAbstractClassObjectFromTs"));
}

TEST_F(EtsGenericTsToEtsTest, check_generic_function_type_any_string)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericFunctionTypeAnyString"));
}

TEST_F(EtsGenericTsToEtsTest, check_generic_function_type_any_int)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericFunctionTypeAnyInt"));
}

TEST_F(EtsGenericTsToEtsTest, check_generic_function_type_any_bool)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericFunctionTypeAnyBool"));
}
// NOTE (#24570): fix interop tests for tuple
TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_function_tuple_type)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericFunctionTupleType"));
}

TEST_F(EtsGenericTsToEtsTest, DISABLED_check_generic_explicitly_declared_type)
{
    ASSERT_EQ(true, CallEtsFunction<bool>(GetPackageName(), "checkGenericExplicitlyDeclaredType"));
}

}  // namespace ark::ets::interop::js::testing
